.github/workflows/ci.yml

```
@@ -23,51 +23,91 @@ permissions:
   packages: read
 
 jobs:
+  dispatch:
+    runs-on: ubuntu-22.04
+    outputs:
+      rust: ${{ steps.need-check.outputs.rust }}
+      python: ${{ steps.need-check.outputs.python }}
+      web: ${{ steps.need-check.outputs.web }}
+    steps:
+      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # pin v3.5.2
+        timeout-minutes: 5
+
+      - uses: dorny/paths-filter@4512585405083f25c027a35db413c2b3b9006d50  # pin v2.11.1
+        id: changes
+        with:
+          list-files: shell
+          filters: .github/filters/ci.yml
+        timeout-minutes: 5
+
+      - name: Determine which workflows need to be run
+        id: need-check
+        run: |
+          set -eux -o pipefail
+          for check in $(env | grep -o 'NEED_CHECK_\w*' ); do
+            env_value=$(eval "echo \$$check")
+            echo "${check#NEED_CHECK_}=$env_value"
+          done | tee -a $GITHUB_OUTPUT
+        env:
+          NEED_CHECK_rust: ${{ github.ref == 'refs/heads/master' || steps.changes.outputs.python-jobs == 'true' }}
+          NEED_CHECK_web: ${{ github.ref == 'refs/heads/master' || steps.changes.outputs.web-jobs == 'true' }}
+          NEED_CHECK_python: ${{ github.ref == 'refs/heads/master' || steps.changes.outputs.rust-jobs == 'true' }}
+
   # Github PR merging is configured to only require this job to pass
   ci-is-happy:
     name: ⭐ CI is happy ⭐
     needs:
-      - check-quality-assurance
-      - test-python-matrix
-      - test-rust-matrix
-      - test-web-app
-      - check-spelling
+      - dispatch
+      - quality-assurance
+      - python
+      - rust
+      - web
+      - spelling
     runs-on: ubuntu-latest
     if: always()
     # Just a fail-safe timeout, see the fine grain per-task timeout instead
     timeout-minutes: 2
     steps:
       # The Needs context value contains only:
       # - the final state a jobs (if it fails or not)
-      # - it's output (actually none of our jobs are configuring outputs variable)
+      # - it's output
       #
       # https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context
-      - name: Debug the needs context values
+      - name: Requirement fullfiled
+        run: |
+          #!python3
+          import os
+          import json
+
+          data = json.loads(os.environ["NEEDS"])
+
+          print(json.dumps(data, indent=4))
```

Print for debug? Otherwise it might be nice to print something before like 

```suggestion
          print("NEEDS:")
          print(json.dumps(data, indent=4))
```

---

It's for debug